bittensor/utils/__init__.py

```diff
@@ -102,7 +106,126 @@ def solve_for_difficulty( block_hash, difficulty ):
+        self.stopEvent = stopEvent
+        self.limit = limit
+
+    def run(self):
+        block_number: int
+        block_bytes: bytes
+        block_difficulty: int
+        nonce_limit = int(math.pow(2,64)) - 1
+
+        # Start at random nonce
+        nonce_start = self.update_interval * self.proc_num + random.randint( 0, nonce_limit )
+        nonce_end = nonce_start + self.update_interval
+        while not self.stopEvent.is_set():
+            if self.newBlockEvent.is_set():
+                with self.check_block:
+                    block_number = self.curr_block_num.value
+                    block_bytes = bytes(self.curr_block)
+                    block_difficulty = int(self.curr_diff[0] >> 32 | self.curr_diff[1])
+
+                self.newBlockEvent.clear()
+                # reset nonces to start from random point
+                nonce_start = self.update_interval * self.proc_num + random.randint( 0, nonce_limit )
+                nonce_end = nonce_start + self.update_interval
+                
+            # Do a block of nonces
+            solution, time = solve_for_nonce_block(self, nonce_start, nonce_end, block_bytes, block_difficulty, self.limit, block_number)
+            if solution is not None:
+                self.solution_queue.put(solution)
+
+            # Send time
+            self.time_queue.put_nowait(time)
+                
+            nonce_start += self.update_interval * self.num_proc
+            nonce_end += self.update_interval * self.num_proc
+
+
+def solve_for_nonce_block(solver: Solver, nonce_start: int, nonce_end: int, block_bytes: bytes, difficulty: int, limit: int, block_number: int) -> Tuple[Optional[POWSolution], int]:
+    best_local = float('inf')
+    best_seal_local = [0]*32
+    start = time.time()
+    for nonce in range(nonce_start, nonce_end):
+        # Create seal.
+        nonce_bytes = binascii.hexlify(nonce.to_bytes(8, 'little'))
+        pre_seal = nonce_bytes + block_bytes
+        seal_sh256 = hashlib.sha256( bytearray(hex_bytes_to_u8_list(pre_seal)) ).digest()
+        kec = keccak.new(digest_bits=256)
+        seal = kec.update( seal_sh256 ).digest()
+        seal_number = int.from_bytes(seal, "big")
+
+        # Check if seal meets difficulty
+        product = seal_number * difficulty
+        if product < limit:
+            print(f"{solver.proc_num} found a solution: {nonce}, {block_number}, {str(block_bytes)}, {str(seal)}, {difficulty}")
+            # Found a solution, save it.
+            return POWSolution(nonce, block_number, difficulty, seal), time.time() - start
+
+        if (product - limit) < best_local: 
+            best_local = product - limit
+            best_seal_local = seal
+
+    # Send best solution to best queue.
+    solver.best_queue.put((best_local, best_seal_local))
+    return None, time.time() - start
+
+
+def solve_for_difficulty_fast( subtensor, wallet, num_processes: int = None, update_interval: int = 50_000 ) -> Optional[POWSolution]:
```

Comment:
Probably better form if we have `update_interval: Optional[int] = None` since we override to the default value below anyways. (And if we change it from 50k, then it only changes in one spot)

Reply:
Added